### ISSUE TITLE
REPP: bulk change - tech contact replacement

### ISSUE THREAD

**Issue by vohmar** [internetee/registry#662]
Case: registrar has taken over (transferred) bulk of domains from another registrar. Now needs to replace previous registrar's tech contact data with its own.

Query replaces specified tech contact of the domains registered using the registrar with another specified tech contact - input contact object code of tech contact to be replaced and object code of the contact that replaces previously specified contact

Bulk change form for tech contact change should be added to registrars' portal. Confirmation must be asked from user before the query is run ("are you sure?" message)

**Comment by artur-beljajev** [comment]
What fields should be in that form? Can you give some example? It would be good to see user story as well.

**Comment by vohmar** [comment]
Lets go for simplest solution possible - bulk transfer that replaces specific tech contact with another existing contact.

**Case/story1:** Registrar has transferred 100 domains from another registrar using bulk transfer REPP command. Most of the transferred domains have previous registrar's DNS admin as technical contact. Registrar logs into the portal for registrars, navigates to bulk tech contact replacement form, sees form with two fields - one for contact id to be replaced (contact1) and the other for contact id to replace the contact with (contact2) - form aids user with auto suggestion functionality, similarly to domain create/update form) After pressing update all the domains of that registrar get domain update to remove contact1 and add contact2 where contact1 is present.

**Comment by vohmar** [comment]
blocked by https://github.com/internetee/registry/issues/746

**Comment by artur-beljajev** [comment]
Could we implement this as a part of transfer operation?

**Comment by vohmar** [comment]
* replacing nameservers and tech contacts on transfer is nice to have idea - needs a bit work though before implementing as registrar might not want to make the change for all the domains being transferred
* registrars can only change data directly associated with them without special consent from registrant. This is limited to tech contacts, nameservers and dnssec keys. Although registrars are sometimes also used as admin contacts. But for now we'll stick only to tech contacts
* code and name is to help user - I have not spotted this type of lag on test nor production system nor have we had any complaints from registrars. But if there is room for improvement to fix the performance  but without loosing the functionality we should discuss it.

**Comment by artur-beljajev** [comment]
> replacing nameservers and tech contacts on transfer is nice to have idea - needs a bit work though before implementing as registrar might not want to make the change for all the domains being transferred

Let's make it optional then!

**Comment by vohmar** [comment]
* We'll handle any changes during transfer in separate ticket
* Yes take into account that there might be other options for bulk changes in addition to domain transfer, nemeserver and tech contact changes

**Comment by artur-beljajev** [comment]
Why confirmation is needed? If you have already reached that form and submitted it, it's very unlikely you did it by accident.

**Comment by vohmar** [comment]
* repp returns empty response 204 on success. Please return list of domains that had tech contacts changed (/admin/repp_logs/569). Improve also the message returned to user in the portal for registrars with the list of domains.
* was able to change tech contacts on domain with delete candidate status - deletecandidate status means that domain is waiting for deletion and nothing can be done to it any more except removing delete candidate status in admin interface.
* Please rename replace domain contacts to replace tech contacts - it is currently too vague and leaves an impression like other types of contacts can or will be updated as well (/registrar/domain_contacts)
* Please move bulk change functionality under one button in domains view (/registrar/domains) ie 'bulk change' and create a tab based logic like there is with domain transfer (/registrar/domain_transfers/new). Add three tabs - transfer, nameserver change, tech contact change

**Comment by artur-beljajev** [comment]
> Improve also the message returned to user in the portal for registrars with the list of domains.

Are you sure you want them all to be listed (and that it will be really useful)? Is there anybody who will study the entire list? As I understand, the amount of domains will be quite high given the nature of the feature.

**Comment by artur-beljajev** [comment]
> Please move bulk change functionality under one button in domains view (/registrar/domains) ie 'bulk change' and create a tab based logic like there is with domain transfer (/registrar/domain_transfers/new). Add three tabs - transfer, nameserver change, tech contact change

You mean move "Replace nameserver" and "Replace technical contacts" forms under tabs on "Domain transfer" page? 

- Why 3 tabs, if "Domain transfer" page already has 2?
- How does "bulk replace" feature connected to domain transfer? As I see it, I can freely replace some legacy tech contacts with the new ones without doing any transfer operation, so perhaps it's better to just unify that new 2 forms into single page and put them under tabs there?

**Comment by vohmar** [comment]
* It is important for registrar to be able to get overview of the domains affected - the list can be long but does not always have to be. This seems to be the only source for registrar to get that information - so yes REPP has to return the list of domains affected
* Create separate button in registrars' portal domain list view (/registrar/domains) aside to the existing 'new domain' and 'transfer' buttons called 'bulk change'. The tabbed view is opened when user presses the buttons similarly to transfer button functionality. The view has three tabs - transfer, nameserver change, tech contact change.

**Comment by artur-beljajev** [comment]
> Create separate button in registrars' portal domain list view (/registrar/domains) aside to the existing 'new domain' and 'transfer' buttons called 'bulk change'. The tabbed view is opened when user presses the buttons similarly to transfer button functionality. The view has three tabs - transfer, nameserver change, tech contact change.

"Domain transfer" page already has 2 tabs (http://prntscr.com/iyxmj4). I don't get why "domain transfer" (single domain transfer, not in batch) should go under "bulk replace".

**Comment by vohmar** [comment]
* Please add a copy of bulk transfer under bulk change view as we discussed.
* Please replace the the predecessor and successor terms with "current contact id" and "new contact id" respectively.

**Comment by artur-beljajev** [comment]
> Please add a copy of bulk transfer under bulk change view as we discussed.
Clarify this. I would like to see some mockup.

Please collect _all_ fixes you want to be done in this ticket.

**Comment by vohmar** [comment]
in addition to previously described fixes please also add explanatory messages to bulk change tabs
* tech contact change: "Replace technical contact specified in 'current contact id' with the one in 'new contact id' on any domain registered under this registrar"
* nameserver change: "Replace nameserver specified in the 'old hostname' with the one in 'new hostname' with optional IPv4 and IPv6 addresses on any domain registered under this registrar"
* transfer: "Transfer domains in the csv file with correct transfer code to this registrar"

currently there are no other fix requests.

**Comment by artur-beljajev** [comment]
I added front-end text field trimmer which should remove leading and trailing spaces in all text fields. Would be good to verify this part.

**Comment by artur-beljajev** [comment]
1. Done
2. Done. `current_contact_id` and `new_contact_id` are definitely better.
